File: app/src/main/java/com/electrosim/Container.java
package com.electrosim;

import java.util.ArrayList;
import java.util.List;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.widget.Button;

public class Container extends SurfaceView implements SurfaceHolder.Callback {

	private ViewThread mThread;
	protected ArrayList<ElementInterface> mElements = new ArrayList<ElementInterface>();
	ElementInterface currentElement = null;
	float currentPointerX, currentPointerY;
	LinkPointInterface tempLinkPoint;
	ConnectorInterface newConnector;
	ConnectorFactory connectorFactory;
	protected List<ConnectorInterface> mConnectors = new ArrayList<ConnectorInterface>();
	Paint paint = new Paint();
	
	long timetracker1;
	long timetracker2;
	boolean delflag;
	
	
	
	

	public Container(Context context) {
		super(context);
		getHolder().addCallback(this);
		mThread = new ViewThread(this);
		paint.setColor(Color.RED);
		paint.setStyle(Paint.Style.STROKE);
	}

	public Container(Context context, AttributeSet attrs) {
		super(context, attrs);
		getHolder().addCallback(this);
		mThread = new ViewThread(this);
		paint.setColor(Color.RED);
		paint.setStyle(Paint.Style.STROKE);

	}

	public void insertElement(ElementInterface element){
		
		synchronized (mElements) {
			mElements.add(element);
		}
	}
	
	public void doDraw(Canvas canvas) {
		canvas.drawColor(Color.BLACK);
		synchronized (mElements) {
			for (ConnectorInterface connector : mConnectors) {
				connector.draw(canvas);
			}

			for (ElementInterface element : mElements) {
				element.doDraw(canvas);
			}

			if (tempLinkPoint != null)
				canvas.drawCircle(tempLinkPoint.getX(), tempLinkPoint.getY(),
						20, paint);

			if (newConnector != null) {
				newConnector.draw(canvas, currentPointerX, currentPointerY);
			}
		}
	}

	@Override
	public void surfaceChanged(SurfaceHolder holder, int format, int width,
			int height) {
	}

	@Override
	public void surfaceCreated(SurfaceHolder holder) {
		if (!mThread.isAlive()) {
			// mElements.add(new Element(getResources(), 30, 10));
			mThread = new ViewThread(this);
			mThread.setRunning(true);
			mThread.start();
		}
	}

	@Override
	public void surfaceDestroyed(SurfaceHolder holder) {
		if (mThread.isAlive()) {
			mThread.setRunning(false);
		}
	}

	private final ElementInterface getCurrentElement(float x, float y) {
		if (currentElement != null && currentElement.insideRectCheck(x, y))
			return currentElement;

		for (ElementInterface ele : mElements) {
			if (ele.insideRectCheck(x, y)) {
				return ele;
			}
		}
		return null;
	}
	
	private final ConnectorInterface getCurrentConnector(float x, float y) {
	
		for (ConnectorInterface c : mConnectors) {
			if (c.pointOnLine(x, y)) {
				return c;
			}
		}
		return null;
	}

	@Override
	public boolean onTouchEvent(MotionEvent event) {
		Log.d("TEST1",
				"E - " + event.toString() + " :: " + event.getEdgeFlags());
		
		
		
		synchronized (mElements) {

			
			// if(event.getAction()==MotionEvent.ACTION_DOWN &&
			// insideRectCheck(event.getX(), event.getY())){
			// mousedown=true;
			// }else if(event.getAction()==MotionEvent.ACTION_UP && mousedown){
			// mousedown=false;
			// }else
			if (event.getAction() == MotionEvent.ACTION_MOVE &&  !isDelflag() ) {
				// Element e = getCurrentElement(event.getX(), event.getY());
				// if(e!=currentElement && currentElement!=null){
				// currentElement.mouseOut(event);
				// }
				// currentElement=e;
				if (currentElement != null) {

					float deltaX = event.getX() - currentPointerX;
					float deltaY = event.getY() - currentPointerY;
					currentElement.translate(deltaX, deltaY);

				} else if (newConnector != null) {
					ElementInterface e = getCurrentElement(event.getX(),
							event.getY());
					if (e != null) {
						tempLinkPoint = e.select(event);
					} else
						tempLinkPoint = null;
				}

				// if(mousedown){
				// x = (int) event.getX();
				// y = (int) event.getY();
				// }
				// if(currentElement!=null){
				// Log.d("HOVER","INSIDE RECT");
				// currentElement.mouseMove(event);
				// }
				currentPointerX = event.getX();
				currentPointerY = event.getY();
			
			} else if (event.getAction() == MotionEvent.ACTION_DOWN) {
				ElementInterface e = getCurrentElement(event.getX(),
						event.getY());
				if(isDelflag()){
					if(e!=null){
						
					}else{
						ConnectorInterface c = getCurrentConnector(event.getX(),
								event.getY());
						if(c!=null){
							
							c.getLinkPointA().setConnector(null);
							c.getLinkPointB().setConnector(null);
							mConnectors.remove(c);
						}
					}
				}else{
					if (e != currentElement && currentElement != null) {
						currentElement.deSelect(event);
						if (e != null) {
							currentElement = e;
							tempLinkPoint = currentElement.select(event);
	
						} else
							currentElement = null;
					} else {
						if (currentElement == null && e != null) {
							currentElement = e;
							tempLinkPoint = currentElement.select(event);
						}
					}
	
					if (tempLinkPoint != null) {
						
						//changes1
						currentElement = null;
						if (newConnector == null) {
							// &
							 timetracker1 = System.currentTimeMillis();
	
							//newConnector = new Connector();
						
							newConnector = connectorFactory.createConnector();
							//$ if it is true
						
							//changes1
							if(newConnector.setLinkPointA(tempLinkPoint))
							
							//then do these two lines
							
						//	newConnector.setLinkPointA(tempLinkPoint);
							{
								tempLinkPoint.setConnector(newConnector);
							
								tempLinkPoint = null;
						
							}
						//if $ is false then make & null
			
							//changes1	
								
							else newConnector = null;
						}
	
					}
	
					// if((event.getEdgeFlags() & MotionEvent.EDGE_BOTTOM)!=0)
					//				if (event.getY() > 630)
						//mElements.add(new Element(getResources(), (int) event
							//	.getX(), (int) event.getY()));
				}
				currentPointerX = event.getX();
				currentPointerY = event.getY();
			} else if (event.getAction() == MotionEvent.ACTION_UP &&  !isDelflag()) {
				if (tempLinkPoint != null) {
					if (newConnector != null) {

						if(newConnector.setLinkPointB(tempLinkPoint))
						
						{
						//newConnector.setLinkPointB(tempLinkPoint);	
						tempLinkPoint.setConnector(newConnector);
						mConnectors.add(newConnector);
						newConnector = null;
						}
						}
					tempLinkPoint = null;
				} else if (newConnector != null){
					
					
					 timetracker2 = System.currentTimeMillis();
					
					if ((timetracker2-timetracker1) > 5000){
						
						//newConnector.setLinkPointA(lp)
						newConnector.getLinkPointA().setConnector(null);
						newConnector =null;
						
					//use time diffrer here
					}else
					
					{	
					 timetracker1 = timetracker2;
						newConnector.addLinkBreakpoint(new LineBreakpoint(
							event.getX(), event.getY(), newConnector));
				
					
					}	
						
					
					}
			}
			// if currentconn != null and templinkpoint ==null

		}
		
		return true;
	}

	public ConnectorFactory getConnectorFactory() {
		return connectorFactory;
	}

	public void setConnectorFactory(ConnectorFactory connectorFactory) {
		this.connectorFactory = connectorFactory;
	}

	public boolean isDelflag() {
		return delflag;
	}

	public void setDelflag(boolean delflag) {
		this.delflag = delflag;
	}
}

File: app/src/main/java/com/electrosim/Container.kt
package com.electrosim

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.SurfaceHolder
import android.view.SurfaceView
import java.util.ArrayList

open class Container : SurfaceView, SurfaceHolder.Callback {
    private var mThread: ViewThread
    protected val mElements: ArrayList<ElementInterface> = ArrayList()
    private var currentElement: ElementInterface? = null
    private var currentPointerX: Float = 0f
    private var currentPointerY: Float = 0f
    private var tempLinkPoint: LinkPointInterface? = null
    private var newConnector: ConnectorInterface? = null
    var connectorFactory: ConnectorFactory? = null
    protected val mConnectors: MutableList<ConnectorInterface> = ArrayList()
    private val paint: Paint = Paint()
    
    private var timetracker1: Long = 0
    private var timetracker2: Long = 0
    var delflag: Boolean = false
    
    constructor(context: Context) : super(context) {
        holder.addCallback(this)
        mThread = ViewThread(this)
        paint.color = Color.RED
        paint.style = Paint.Style.STROKE
    }
    
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        holder.addCallback(this)
        mThread = ViewThread(this)
        paint.color = Color.RED
        paint.style = Paint.Style.STROKE
    }
    
    fun insertElement(element: ElementInterface) {
        synchronized(mElements) {
            mElements.add(element)
        }
    }
    
    fun doDraw(canvas: Canvas) {
        canvas.drawColor(Color.BLACK)
        synchronized(mElements) {
            for (connector in mConnectors) {
                connector.draw(canvas)
            }
            
            for (element in mElements) {
                element.doDraw(canvas)
            }
            
            tempLinkPoint?.let {
                canvas.drawCircle(it.getX(), it.getY(), 20f, paint)
            }
            
            newConnector?.let {
                it.draw(canvas, currentPointerX, currentPointerY)
            }
        }
    }
    
    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        // Empty implementation
    }
    
    override fun surfaceCreated(holder: SurfaceHolder) {
        if (!mThread.isAlive) {
            mThread = ViewThread(this)
            mThread.setRunning(true)
            mThread.start()
        }
    }
    
    override fun surfaceDestroyed(holder: SurfaceHolder) {
        if (mThread.isAlive) {
            mThread.setRunning(false)
        }
    }
    
    private fun getCurrentElement(x: Float, y: Float): ElementInterface? {
        if (currentElement?.insideRectCheck(x, y) == true) {
            return currentElement
        }
        
        for (ele in mElements) {
            if (ele.insideRectCheck(x, y)) {
                return ele
            }
        }
        return null
    }
    
    private fun getCurrentConnector(x: Float, y: Float): ConnectorInterface? {
        for (c in mConnectors) {
            if (c.pointOnLine(x, y)) {
                return c
            }
        }
        return null
    }
    
    override fun onTouchEvent(event: MotionEvent): Boolean {
        Log.d("TEST1", "E - $event :: ${event.edgeFlags}")
        
        synchronized(mElements) {
            when (event.action) {
                MotionEvent.ACTION_MOVE -> {
                    if (!delflag) {
                        if (currentElement != null) {
                            val deltaX = event.x - currentPointerX
                            val deltaY = event.y - currentPointerY
                            currentElement?.translate(deltaX, deltaY)
                        } else if (newConnector != null) {
                            val e = getCurrentElement(event.x, event.y)
                            tempLinkPoint = if (e != null) {
                                e.select(event)
                            } else {
                                null
                            }
                        }
                        
                        currentPointerX = event.x
                        currentPointerY = event.y
                    }
                }
                
                MotionEvent.ACTION_DOWN -> {
                    val e = getCurrentElement(event.x, event.y)
                    
                    if (delflag) {
                        if (e == null) {
                            val c = getCurrentConnector(event.x, event.y)
                            c?.let {
                                it.getLinkPointA().setConnector(null)
                                it.getLinkPointB().setConnector(null)
                                mConnectors.remove(it)
                            }
                        }
                    } else {
                        if (e != currentElement && currentElement != null) {
                            currentElement?.deSelect(event)
                            if (e != null) {
                                currentElement = e
                                tempLinkPoint = currentElement?.select(event)
                            } else {
                                currentElement = null
                            }
                        } else if (currentElement == null && e != null) {
                            currentElement = e
                            tempLinkPoint = currentElement?.select(event)
                        }
                        
                        tempLinkPoint?.let {
                            currentElement = null
                            if (newConnector == null) {
                                timetracker1 = System.currentTimeMillis()
                                
                                connectorFactory?.let { factory ->
                                    newConnector = factory.createConnector()
                                    
                                    newConnector?.let { connector ->
                                        if (connector.setLinkPointA(it)) {
                                            it.setConnector(connector)
                                            tempLinkPoint = null
                                        } else {
                                            newConnector = null
                                        }
                                    }
                                }
                            }
                        }
                    }
                    
                    currentPointerX = event.x
                    currentPointerY = event.y
                }
                
                MotionEvent.ACTION_UP -> {
                    if (!delflag) {
                        if (tempLinkPoint != null && newConnector != null) {
                            val lp = tempLinkPoint
                            val connector = newConnector
                            if (lp != null && connector != null && connector.setLinkPointB(lp)) {
                                lp.setConnector(connector)
                                mConnectors.add(connector)
                                newConnector = null
                            }
                            tempLinkPoint = null
                        } else if (newConnector != null) {
                            val connector = newConnector
                            if (connector != null) {
                                timetracker2 = System.currentTimeMillis()
                                
                                if ((timetracker2 - timetracker1) > 5000) {
                                    connector.getLinkPointA().setConnector(null)
                                    newConnector = null
                                } else {
                                    timetracker1 = timetracker2
                                    connector.addLinkBreakpoint(LineBreakpoint(event.x, event.y, connector))
                                }
                            }
                        }
                    }
                }
            }
        }
        
        return true
    }
} 